File: src/j/tests/MicroSet_Benchmark_Remove.java
package j.tests;


import com.google.caliper.BeforeExperiment;
import com.google.caliper.Benchmark;
import com.google.caliper.api.VmOptions;
import com.google.caliper.runner.CaliperMain;
import j.implementation.TestObject;
import j.implementation.microSet.MicroSet;
import j.implementation.microSet.InnerArraySet;

import java.util.HashSet;
import java.util.Random;
import java.util.TreeSet;

@VmOptions("-XX:-TieredCompilation")
public class MicroSet_Benchmark_Remove {

    private static int NUMBER_OF_TEST_OBJECT = 30;

    MicroSet<TestObject> microInnerSet = new MicroSet<>(MicroSet.Use.INNER_SET, NUMBER_OF_TEST_OBJECT);
    MicroSet<TestObject> microArraySet = new MicroSet<>(MicroSet.Use.ARRAY_SET, NUMBER_OF_TEST_OBJECT);
    MicroSet<TestObject> microHashSet = new MicroSet<>(MicroSet.Use.HASH_SET, NUMBER_OF_TEST_OBJECT);
    HashSet<TestObject> hashSet = new HashSet<>();
    TreeSet<TestObject> treeSet = new TreeSet<>();


    TestObject[] testObjects = new TestObject[NUMBER_OF_TEST_OBJECT];
    int[] randomInt = new int[NUMBER_OF_TEST_OBJECT];

    // Half of the testObject will not be contained
    @BeforeExperiment
    void setUp() {
        Random random = new Random();
        for (int i = 0; i < NUMBER_OF_TEST_OBJECT; i++) {
            testObjects[i] = new TestObject(random.nextInt());
        }

        for (int j = 0; j < NUMBER_OF_TEST_OBJECT /2; j++) {
            hashSet.add(testObjects[j]);
            microArraySet.add(testObjects[j]);
            microHashSet.add(testObjects[j]);
            microInnerSet.add(testObjects[j]);
            treeSet.add(testObjects[j]);
        }

        for (int i = 0; i < NUMBER_OF_TEST_OBJECT; i++) {
            randomInt[i] = random.nextInt(NUMBER_OF_TEST_OBJECT);
        }

    }

    @Benchmark
    public void testRemove_HashSet(int reps) {
        for (int i = 0; i < reps; i++) {
            for (int j = 0; j < NUMBER_OF_TEST_OBJECT/4; j++) {
                hashSet.remove(testObjects[randomInt[j]]);
            }
        }
    }

    @Benchmark
    public void testRemove_TreeSet(int reps) {
        for (int i = 0; i < reps; i++) {
            for (int j = 0; j < NUMBER_OF_TEST_OBJECT/4; j++) {
                treeSet.remove(testObjects[randomInt[j]]);
            }
        }
    }

    @Benchmark
    public void testRemove_MicroArraySet(int reps) {
        for (int i = 0; i < reps; i++) {
            for (int j = 0; j < NUMBER_OF_TEST_OBJECT; j++) {
                microArraySet.remove(testObjects[randomInt[j]]);
            }
        }
    }

    @Benchmark
    public void testRemove_MicroHashSet(int reps) {
        for (int i = 0; i < reps; i++) {
            for (int j = 0; j < NUMBER_OF_TEST_OBJECT; j++) {
                microHashSet.remove(testObjects[randomInt[j]]);
            }
        }
    }
    @Benchmark
    public void testRemove_MicroInnerSet(int reps) {
        for (int i = 0; i < reps; i++) {
            for (int j = 0; j < NUMBER_OF_TEST_OBJECT; j++) {
                microInnerSet.remove(testObjects[randomInt[j]]);
            }
        }
    }

    public static void main(String[] args) {
        args = new String[]{ "-i", "runtime", "-r", "REMOVE scale=" + NUMBER_OF_TEST_OBJECT + ", object_number=" + NUMBER_OF_TEST_OBJECT};
        CaliperMain.main(MicroSet_Benchmark_Remove.class, args);
    }
}

File: src/scala/tests/MicroSet_Benchmark_Remove.scala
package scala.tests

import java.util.{HashSet, Random, TreeSet}

import com.google.caliper.api.VmOptions
import com.google.caliper.runner.CaliperMain
import com.google.caliper.{BeforeExperiment, Benchmark}

import scala.implementation.{MicroSet, Use}


/**
  * Created by pxl on 03/04/17.
  */

@VmOptions (Array ("-XX:-TieredCompilation") )
class MicroSet_Benchmark_Remove {

  val microInnerSet = new MicroSet[TestObject]()
  MicroSet.use = Use.ARRAY_SET
  val microArraySet = new MicroSet[TestObject]()
  MicroSet.use = Use.HASH_SET
  val microHAshSet = new MicroSet[TestObject]()

  val hashSet = new HashSet[TestObject]
  val treeSet = new TreeSet[TestObject]

  var tabTestObject: Array[TestObject] = Array.ofDim(runnerRemove.NUMBER_OF_TEST_OBJECT)
  var tabRandomInt: Array[Int] = Array.ofDim(runnerRemove.NUMBER_OF_TEST_OBJECT)

  @BeforeExperiment()
  def setUp() = {
    val rand: Random = new Random
    for (i <- 0 to runnerRemove.NUMBER_OF_TEST_OBJECT) tabTestObject(i) = TestObject(rand.nextInt())
    for (i <- 0 to runnerRemove.NUMBER_OF_TEST_OBJECT / 2) {
      microArraySet.add(tabTestObject(i))
      microHAshSet.add(tabTestObject(i))
      microInnerSet.add(tabTestObject(i))

      hashSet.add(tabTestObject(i))
      treeSet.add(tabTestObject(i))
    }
    for (i <- 0 until runnerRemove.NUMBER_OF_TEST_OBJECT) tabRandomInt(i) = rand.nextInt(runnerRemove.NUMBER_OF_TEST_OBJECT)
  }

  @Benchmark
  def testRemove_TreeSet(reps: Int) = {
    for (i <- 0 to reps) {
      for (j <- 0 until runnerRemove.NUMBER_OF_TEST_OBJECT / 4) treeSet.remove(tabTestObject(tabRandomInt(j)))
    }
  }

  @Benchmark
  def testRemove_HashSet(reps: Int) = {
    for (i <- 0 to reps) {
      for (j <- 0 until runnerRemove.NUMBER_OF_TEST_OBJECT / 4) hashSet.remove(tabTestObject(tabRandomInt(j)))
    }
  }

  @Benchmark
  def testRemove_InnerSet(reps: Int) = {
    for (i <- 0 to reps) {
      for (j <- 0 until runnerRemove.NUMBER_OF_TEST_OBJECT) microInnerSet.remove(tabTestObject(tabRandomInt(j)))
    }
  }

  @Benchmark
  def testRemove_InnerArraySet(reps: Int) = {
    for (i <- 0 to reps) {
      for (j <- 0 until runnerRemove.NUMBER_OF_TEST_OBJECT) microArraySet.remove(tabTestObject(tabRandomInt(j)))
    }
  }

  @Benchmark
  def testRemove_InnerHashSet(reps: Int) = {
    for (i <- 0 to reps) {
      for (j <- 0 until runnerRemove.NUMBER_OF_TEST_OBJECT) microHAshSet.remove(tabTestObject(tabRandomInt(j)))
    }
  }
}

object runnerRemove
{
  val NUMBER_OF_TEST_OBJECT: Int = 30
  def main(args: Array[String]): Unit = {
    val args = Array[String]("-i", "runtime", "-r", "ADD scale=" + NUMBER_OF_TEST_OBJECT + ", object_number=" + NUMBER_OF_TEST_OBJECT)
    CaliperMain.main(classOf[MicroSet_Benchmark_Remove], args)
  }

}
